src/controls/experimental/CameraControls.ts

```diff
@@ -0,0 +1,1176 @@
+
+    this.rotateStart = new Vector2()
+    this.rotateEnd = new Vector2()
+    this.rotateDelta = new Vector2()
+
+    this.panStart = new Vector2()
+    this.panEnd = new Vector2()
+    this.panDelta = new Vector2()
+
+    this.dollyStart = new Vector2()
+    this.dollyEnd = new Vector2()
+    this.dollyDelta = new Vector2()
+
+    this.domElement.addEventListener('contextmenu', this.onContextMenu, false)
+
+    this.domElement.addEventListener('mousedown', this.onMouseDown, false)
+    this.domElement.addEventListener('wheel', this.onMouseWheel, false)
+
+    this.domElement.addEventListener('touchstart', this.onTouchStart, false)
+    this.domElement.addEventListener('touchend', this.onTouchEnd, false)
+    this.domElement.addEventListener('touchmove', this.onTouchMove, false)
+
+    this.domElement.addEventListener('keydown', this.onKeyDown, false)
+
+    // make sure element can receive keys.
+
+    if (this.domElement.tabIndex === -1) {
+      this.domElement.tabIndex = 0
+    }
+
+    // force an update at start
+
+    this.object.lookAt(this.target)
+    this.update()
+    this.saveState()
+  }
+
+  getPolarAngle = () => this.spherical.phi
+
+  getAzimuthalAngle = () => this.spherical.theta
+
+  saveState = () => {
+    this.target0.copy(this.target)
+    this.position0.copy(this.object.position)
+    this.quaternion0.copy(this.object.quaternion)
+    this.zoom0 = this.object.zoom
+  }
+
+  reset = () => {
+    this.target.copy(this.target0)
+    this.object.position.copy(this.position0)
+    this.object.quaternion.copy(this.quaternion0)
+    this.object.zoom = this.zoom0
+
+    this.object.updateProjectionMatrix()
+    this.dispatchEvent(this.changeEvent)
+
+    this.update()
+
+    this.state = STATE.NONE
+  }
+
+  dispose = () => {
+    this.domElement.removeEventListener('contextmenu', this.onContextMenu, false)
+    this.domElement.removeEventListener('mousedown', this.onMouseDown, false)
+    this.domElement.removeEventListener('wheel', this.onMouseWheel, false)
+
+    this.domElement.removeEventListener('touchstart', this.onTouchStart, false)
+    this.domElement.removeEventListener('touchend', this.onTouchEnd, false)
+    this.domElement.removeEventListener('touchmove', this.onTouchMove, false)
+
+    document.removeEventListener('mousemove', this.onMouseMove, false)
+    document.removeEventListener('mouseup', this.onMouseUp, false)
+
+    this.domElement.removeEventListener('keydown', this.onKeyDown, false)
+
+    //this.dispatchEvent( { type: 'dispose' } ); // should this be added here?
+  }
+
+  private update = (() => {
+    const offset = new Vector3()
+
+    // so camera.up is the orbit axis
+    const quat = new Quaternion().setFromUnitVectors(this.object.up, new Vector3(0, 1, 0))
+    const quatInverse = quat.clone().invert()
+
+    const lastPosition = new Vector3()
+    const lastQuaternion = new Quaternion()
+
+    const q = new Quaternion()
+    const vec = new Vector3()
+
+    return () => {
```

Comment:
I think we should be moving the above props to private (likes take advantage of TS) and then we can use the function below as default, makes it cleaner, what do you think @giulioz?

Reply:
Yeah we could, even though it can also be done later.
We also need to make everything consistent with the other controls

Reply:
It'd be in line with what I did for other classes relating to `drei`

Reply:
moved the values out to be private props + made the methods contain just the inner functions :smiley: